### PR TITLE
Fixes to get orcania building with MSVC

### LINKED ISSUE #29: ssize_t is not available on MSVC
Building orcania 2.3.1 with Visual Studio fails because `ssize_t` is not known.

This line fails to compile:
https://github.com/babelouest/orcania/blob/c66e9a44b218de9f9187507d8037ef7d7dade855/src/orcania.c#L347
with the following error:
```
C:\Users\maarten\.conan\data\orcania\2.3.1\_\_\build\3fb49604f9c2f729b85ba3115852006824e72cab\src\src\
orcania.c(347,11): error C2065: 'ssize_t': undeclared identifier [C:\Users\maarten\.conan\data\orcania
\2.3.1\_\_\build\3fb49604f9c2f729b85ba3115852006
```

### PR DESCRIPTION
**Pull request by madebr**
Hello, I'm in the process with creating a conan recipe for ulfius, which has orcania as one of its dependencies.
By applying these patches, I am able to build this library with Visual Studio 2019.

- `ssize_t` is not available with Visual Studio. So I applied [this suggestion](https://www.scivision.dev/ssize_t-visual-studio-posix/)
- conan recipes usually have an option to package shared/static libraries. To avoid having to build the shared library, when only wanting to package the shared library, this pr adds a `BUILD_SHARED` cmake option.
  I have also fixed building a dll by setting the [`WINDOWS_EXPORT_ALL_SYMBOLS`](https://cmake.org/cmake/help/latest/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html) property.
  I was careful to not change the default behavior of the cmake script.
- `base64url.exe` also failed to build because the size of an array must be a constant.

Fixes #29 
Used in https://github.com/conan-io/conan-center-index/pull/13589

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -112,29 +112,47 @@ include_directories(${PROJECT_BINARY_DIR})
 
 # static library
 
+option(BUILD_SHARED "Build shared library." ON)
 option(BUILD_STATIC "Build static library." OFF)
 
+if (NOT BUILD_STATIC AND NOT BUILD_SHARED)
+    message(FATAL_ERROR "BUILD_SHAREDand BUILD_STATIC cannot be both disabled")
```
Comment (babelouest) [inline CMakeLists.txt]: There a typo, a space character missing between `BUILD_SHARED` and `and`

File: CMakeLists.txt [hunk 2]
```diff
@@ -112,29 +112,47 @@ include_directories(${PROJECT_BINARY_DIR})
...
-            VERSION "${LIBRARY_VERSION}"
-            SOVERSION "${LIBRARY_SOVERSION}")
-endif()
-if (WIN32)
-    set_target_properties(orcania PROPERTIES SUFFIX "-${LIBRARY_VERSION_MAJOR}.dll")
+            PUBLIC_HEADER "${INC_DIR}/orcania.h;${PROJECT_BINARY_DIR}/orcania-cfg.h")
+    if (NOT MSVC)
+        set_target_properties(orcania PROPERTIES
+                COMPILE_OPTIONS -Wextra
+                VERSION "${LIBRARY_VERSION}"
```
Comment (babelouest) [inline CMakeLists.txt]: Question: Why is `VERSION` parameter not included in MSVC build?

File: src/orcania.c [hunk 3]
```diff
@@ -38,6 +38,13 @@
 #define strcasecmp _stricmp
 #endif
 
+#ifdef _MSC_VER
+#include <BaseTsd.h>
+typedef SSIZE_T ssize_t;
+#else
+#include <unistd.h>
```
Comment (babelouest) [inline src/orcania.c]: I'd prefer to have `#include <unistd.h>` not here if it's not required

File: tools/base64url/base64url.c [hunk 4]
```diff
@@ -98,12 +105,12 @@ static unsigned char * get_file_content(const char * file_path, size_t * length)
 }
 
 static unsigned char * get_stdin_content(size_t * length) {
-  int size = 100;
-  unsigned char * out = NULL, buffer[size];
+#define SIZE 100
```
Comment (babelouest) [inline tools/base64url/base64url.c]: Usually I put my `#define` statements on top of the file, for better readability

**Comment by babelouest** [comment]
Thanks for the PR @madebr !

I've added a few comments in it, if you don't mind making the change (or challenging them!)

I suppose there should be similar changes in yder and ulfius to make it work for your repo?

File: src/orcania.c (hunk 3 above)
Comment (madebr) [reply src/orcania.c]: I think I mistakenly copy pasted this from `base64url.c`.

File: tools/base64url/base64url.c (hunk 4 above)
Comment (madebr) [reply tools/base64url/base64url.c]: What do you think about the following alternative:
- hard code 100 here,
- and use an `ARRAY_SIZE`-like macro as argument to `read` (or a plain `sizeof`)

File: CMakeLists.txt (hunk 2 above)
Comment (madebr) [reply CMakeLists.txt]: To be honest, I applied the least amount of changes to the cmake script.
`VERSION` is mostly used for versioning `.so` and `.dylib` libraries. For dll's is changing a built-in version.
See [`VERSION`](https://cmake.org/cmake/help/latest/prop_tgt/VERSION.html).

I have added a patch that improves (imho) the cmake script a bit by using targets.
I think it's still compatible with CMake 3.5.

File: CMakeLists.txt (hunk 1 above)
Comment (madebr) [reply CMakeLists.txt]: This is fixed.

**Comment by madebr** [comment]
The 4th patch makes extensive use of CMake targets, instead of putting everything into variables.
So use `Subunit::Subunit`/`Check::Check` as target names for the test libraries.

Create `Orcania::Orcania` for the shared library, and `Orcania:Orcania-static` for the static library.
By using targets, there is no need to always add `${LIBS}` to the linked library list.
The same patch also installs cmake config files:.
This is how an install prefix looks like when building with Visual Studio.
```
prefix
├── bin
│   └── orcania-2.dll
├── cmake
│   ├── OrcaniaConfig.cmake
│   ├── OrcaniaConfigVersion.cmake
│   ├── OrcaniaTargets.cmake
│   └── OrcaniaTargets-debug.cmake
├── include
│   ├── orcania.h
│   └── orcania-cfg.h
└── lib
    ├── orcania.lib
    ├── orcania-static.lib
    └── pkgconfig
        └── liborcania.pc
```

With gcc on Linux, the cmake config files will be installed to `lib/cmake/Orcania`.
I choose capitalized `Orcania` because `ulfius` also contains `FindOrcania.cmake`.

File: tools/base64url/base64url.c (hunk 4 above)
Comment (babelouest) [reply tools/base64url/base64url.c]: It's a nice code to look at, but I'd rather not making it more complex than it should be.

I use a simple `#define` statement in those cases and it simplifies the code and easier to debug.

Comment (madebr) [reply tools/base64url/base64url.c]: Aight. This is the way.

**Comment by babelouest** [comment]
Thanks for the big help with the cmake script!

Can you explain a little bit the purpose of the files installed in the `cmake` directory?
- OrcaniaConfig.cmake
- OrcaniaConfigVersion.cmake
- OrcaniaTargets.cmake
- OrcaniaTargets-debug.cmake

I would like to correctly understand the new features.

**Comment by madebr** [comment]
Sure, these files allow consumers of orcania to use it with the following script.
Note the absence of `find_library`, `find_path` and no need for a custom `FindOrcania.cmake`.

```cmake
cmake_minimum_required(VERSION 3.0)
project(my_program C)
find_package(Orcania 2.3.1 REQUIRED)
add_executable(my_program my_source.c)
target_link_libraries(my_program PRIVATE Orcania::Orcania)
```

- `OrcaniaConfigVersion.cmake`: this checks whether this particular orcania install has a correct version.
It also does a basic test for pointer size (32-bit vs 64-bit).
- `OrcaniaConfig.cmake`: this file creates the actual CMake variables + targets. It can create custom functions etc for your project etc.
Here, it simply imports `OrcaniaTargets.cmake` and defines `ORCANIA_LIBRARIES` and `ORCANIA_INCLUDE_DIRS` variables.
- `OrcaniaTargets.cmake`: this file is generated by doing `install(EXPORT)`. As the name says, it exports targets to be used outside the orcania project.
- `OrcaniaTargets-debug.cmake`: this file is also generated by doing `install(EXPORTS)`. But where `OrcaniaTargets.cmake` creates the targest, this file fills in the location of the libraries.
When you build orcania in release mode, `OrcaniaTargets-release.cmake` is generated instead.
In principle, `OrcaniaTargets-release.cmake` can be installed alongside `OrcaniaTargets-debug.cmake` if Debug and Release libraries don't overlap.

**Comment by babelouest** [comment]
So I could get rid of the custom files `Find[Orcania|Yder|Ulfius|etc.].cmake` ? That's a good news!

This refactor is very much appreciated!

**Comment by madebr** [comment]
Yes, these `FindXXX.cmake` are useful when you don't have control over your dependencies.
For compatibility with older yder/orcania releases, you might want to keep these files in-repo for a while.

Please hold on merging, because I want to test this on Linux too (working on Windows atm). I've marked it as a draft for the time being.

**Comment by babelouest** [comment]
> Please hold on merging, because I want to test this on Linux too (working on Windows atm). I've marked it as a draft for the time being.

Anyway, if you don't mind, I intend to merge not before a few days, until I understand the changes and make sure with my [CI](https://github.com/babelouest/huddersfield) that the rest is fine too.

**Comment by madebr** [comment]
FYI, these cmake config files are completely relocatable.
Which means you can copy an orcania prefix wherever you want, and it will still work.
You just need to add the root of the orcania prefix to `CMAKE_PREFIX_PATH` when configuring your dependency.

**Comment by madebr** [comment]
I verified this on Linux, and used it as a dependency for yder (see https://github.com/babelouest/yder/pull/23)

**Comment by babelouest** [comment]
@madebr , please let me know when you'll have a complete PR, because it's hard to follow-up with your push-forced commits

File: CMakeLists.txt [hunk 5]
```diff
@@ -244,24 +269,47 @@ configure_file(liborcania.pc.in liborcania.pc @ONLY)
...
+install(EXPORT OrcaniaExports DESTINATION "${ORCANIA_INSTALL_CMAKEDIR}"
+        NAMESPACE "Orcania::"
+        FILE "OrcaniaTargets.cmake")
+
+configure_package_config_file(cmake-modules/OrcaniaConfig.cmake.in OrcaniaConfig.cmake
+        INSTALL_DESTINATION "${ORCANIA_INSTALL_CMAKEDIR}")
+write_basic_package_version_file(OrcaniaConfigVersion.cmake
+        COMPATIBILITY AnyNewerVersion)
+
+install(FILES
```
Comment (babelouest) [inline CMakeLists.txt]: IMHO, the `/usr/lib/cmake/Orcania/` files should be installed only in development mode, i.e. when the option `INSTALL_HEADER` is `ON`, don't you think?

Comment (madebr) [reply CMakeLists.txt]: Sounds good. Will do that.

Comment (madebr) [reply CMakeLists.txt]: Fixed in 52735b0

**Comment by madebr** [comment]
> @madebr , please let me know when you'll have a complete PR, because it's hard to follow-up with your push-forced commits

I will stop force pushing then. I think this pr is finished. At the end of the ride, you can squash them.
I could just copy-paste these patches for use at ConanCenter.

**Comment by babelouest** [comment]
> I will stop force pushing then. I think this pr is finished. At the end of the ride, you can squash them. I could just copy-paste these patches for use at ConanCenter.

Awesome! I'll complete the review and my tests and I'll merge this PR and the other 2 then. Thanks again @madebr !

**Comment by babelouest** [comment]
Sorry for the conflicts on your PRs @madebr , I'm still processing your new features, but I'm not ready yet to push them, stay tuned!

**Comment by madebr** [comment]
Do you want me to fix the conflicts?

**Comment by babelouest** [comment]
> Do you want me to fix the conflicts?

Yes if you don't mind, thanks!

**Comment by madebr** [comment]
Friendly ping to ask whether you need any further help?

**Comment by babelouest** [comment]
Sorry, I have long days these times...

I need to take a look at the 3 PRs at the same time, that's why it takes longer time. I'm not forgetting then though.
